sassanh/qnvim
# Use CMake to build plugin + fetch dependencies automatically

## Summary

This change ports the QMake build to CMake - a recommended way to build plugins according to the Qt Documentation:
https://doc.qt.io/qtcreator-extending/first-plugin.html#cmake-project

This also adds a number of useful features, for example, an ability to fetch neovim-qt and qtcreator automatically.

Neovim-qt is downloaded and built with the Message Pack using FetchContent CMake feature. In case of QtCreator only the precompiled binaries are downloaded to save time.

## Test Plan

0. Ensure Neovim is installed and in the PATH
1. Open the CMake project in QtCreator (any version).
2. Once successfully configured, set in run configuration the following:
    - Executable: `%{buildDir}\_deps\qtcreator-src\bin\qtcreator.exe`
    - Command line arguments: `-pluginpath %{buildDir}\lib\qtcreator\plugins`
 3. Build the project and run. This will launch a Qt Creator of compatible version.
 4. Open a file and test that the plugin is working (kinda, but I think the bugs are unrelated to CMake port).

Configurations tested:
- [x] Windows, building the project using MinGW
- [ ] Windows, building the project using MSVC

## TODO

- [ ] Change the documentation on how to build and install
